### ISSUE TITLE
It should show type error when dealing with assigned variables

### ISSUE THREAD

**Issue by anyexinglu** [microsoft/TypeScript#36825]
**Code**

```tsx
import React from 'react'

export interface ColumnProps<T = {}> {
  // 列的 key，用于和 dataSorce 进行一一对应
  key: string
  // 列的名称
  title?: React.ReactNode
  // 渲染函数，可以进行额外操作
  render?: (text: any, record: T, rowIndex: number) => React.ReactNode
}

export function Table<T>(props: { dataSource: T[]; columns: ColumnProps<T>[] }) {
  return <div>tb</div>
}

export const columns1: ColumnProps<{
  goodsId: number
  goodsName: string
  amount: string
  address: JSX.Element
  count: string
  editId: string
  goodsPrice: number
  goodsQuatity: number
}>[] = [
  {
    key: 'name',
    test: 'uyxy'
  }
]

export const columns2 = [
  {
    key: 'name',
    test: 'uyxy'
  }
]

const dataSource = [
  {
    goodsId: 111,
    goodsName: '某商品',
    amount: '18',
    address: (
      <div>
        某地方
        <br />
        某地方
        <br />
        某地方
      </div>
    ),
    count: '已上架',
    editId: '0111',
    goodsPrice: 1523,
    goodsQuatity: 2323
  },
  {
    goodsId: 222,
    goodsName: '某商品',
    amount: '18',
    address: (
      <div>
        某地方
        <br />
        某地方
        <br />
        某地方
      </div>
    ),
    count: '已上架',
    editId: '0222',
    goodsPrice: 1523,
    goodsQuatity: 2323
  }
]

const Demo = (
  <div>
    <Table dataSource={dataSource} columns={columns1} />
    <Table dataSource={dataSource} columns={columns2} />
    <Table
      dataSource={dataSource}
      columns={[
        {
          key: 'name',
          test: 'uyxy'
        }
      ]}
    />
  </div>
)
```

**Expected behavior:**

The last two Table should show Error too as the first Table：

不能将类型“{ key: string; test: string; }”分配给类型“ColumnProps<{ goodsId: number; goodsName: string; amount: string; address: Element; count: string; editId: string; goodsPrice: number; goodsQuatity: number; }>”。

**Actual behavior:**

- In my local IDE, only the first Table and third Table show the error.
![image](https://user-images.githubusercontent.com/8243326/74602095-4ba19b00-50e0-11ea-8a9f-ef6f360378d7.png)

- In the codesandbox demo (with the same version of typescript '3.7.4'), only the first Table shows the error.
![image](https://user-images.githubusercontent.com/8243326/74602100-6542e280-50e0-11ea-9686-cb2897516090.png)

**Playground Link:**  https://codesandbox.io/s/boring-zhukovsky-s3wyn

**Comment by jcalz** [comment]
A more compact example would be useful here.

There's no error with `columns2` because excess properties do not violate an interface.  That is, object types in TypeScript are not "exact"; see #12936.  Excess properties are only considered an error when used in an object literal, see [the documentation](http://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks) or #3823.

The third `Table` should be an error, and that's what I see in the Playground: [link](http://www.typescriptlang.org/play/?jsx=2#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwCwAUMAHYxZQF5ZwBiEEcAN6M4KAFxxUMKCwDm9BgF9GAEyy4ANskpwCAV2b5gEZnABSACjDEwqCYN38JfCAG0AunEUBKCWYDKABoAdACiGlggWKwKADxmjhAAvIKuDsgSZGQANHAARhIAjF7uinAA9AB8NBXlcOzEUCjoWAAeYOpsKiJAA).  I'm not sure why your codesandbox does not show an error, but I don't know that this issue tracker is the right place to debug other IDEs.  If there's a real bug in TypeScript somewhere it would be nice to reproduce it in the Playground specifically.  Can you do this?  Good luck!

**Comment by anyexinglu** [comment]
> A more compact example would be useful here.
> 
> There's no error with `columns2` because excess properties do not violate an interface. That is, object types in TypeScript are not "exact"; see #12936. Excess properties are only considered an error when used in an object literal, see [the documentation](http://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks) or #3823.
> 
> The third `Table` should be an error, and that's what I see in the Playground: [link](http://www.typescriptlang.org/play/?jsx=2#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwCwAUMAHYxZQF5ZwBiEEcAN6M4KAFxxUMKCwDm9BgF9GAEyy4ANskpwCAV2b5gEZnABSACjDEwqCYN38JfCAG0AunEUBKCWYDKABoAdACiGlggWKwKADxmjhAAvIKuDsgSZGQANHAARhIAjF7uinAA9AB8NBXlcOzEUCjoWAAeYOpsKiJAA). I'm not sure why your codesandbox does not show an error, but I don't know that this issue tracker is the right place to debug other IDEs. If there's a real bug in TypeScript somewhere it would be nice to reproduce it in the Playground specifically. Can you do this? Good luck!

@jcalz , codesandbox has some bug, so consider the vscode IDE only.
The explanation of `columns2` -- "excess properties do not violate an interface" is the answer, thx!
